### ISSUE TITLE
Docker Hangs Under Load

### ISSUE THREAD

**Issue by AlecIsaacson** [docker/docker#26995]
I have a small environment with four Docker containers on a single Centos 7.2 host.  If I put the containers under load, Docker hangs.  While it's hung, docker stats shows no values, docker ps works, but docker exec and docker-compose down all hang.

I can avoid the hang in some cases by putting a long delay (30 seconds) between putting load on each container.  In other words, I put load on container 1, wait 30 seconds, put load on container 2, etc, so after 2 minutes, all 4 containers are under load.  However, once we're all under load, if I issue a command against any container with docker exec, I get a hang.

The only fix is to stop Docker (service docker stop) and then restart.

strace against the dockerd process shows this:

Process 7097 attached
wait4(7103,

strace against process 7103 (which is docker-containerd), shows this:

Process 7103 attached
futex(0xed8348, FUTEX_WAIT, 0, NULL

I issued a sigusr1 against the dockerd and here is the trace:

[dockerd-dump.txt](https://github.com/docker/docker/files/499308/dockerd-dump.txt)

**Steps to reproduce the issue:**
1. Start Docker service with docker-compose (or by hand).
2. Put load on the containers (either with real load or by running stress-ng inside of each container)
3. Run a command against one of the containers with docker exec
4. It hangs.

**Describe the results you received:**

Docker became non-responsive. Docker stats shows all lines.

**Describe the results you expected:**

Docker runs as normal and accepts commands.

**Additional information you deem important (e.g. issue happens only occasionally):**

I can make this issue occur on demand.

**Output of `docker version`:**
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:
 OS/Arch:      linux/amd64

```
(paste your output here)
```

**Output of `docker info`:**
Containers: 4
 Running: 4
 Paused: 0
 Stopped: 0
Images: 10
Server Version: 1.12.1
Storage Driver: devicemapper
 Pool Name: docker-thinpool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 2.337 GB
 Data Space Total: 61.2 GB
 Data Space Available: 58.86 GB
 Metadata Space Used: 766 kB
 Metadata Space Total: 641.7 MB
 Metadata Space Available: 641 MB
 Thin Pool Minimum Free Space: 6.119 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.36.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.64 GiB
Name: usildodnimladock1
ID: KZ2H:IWUH:KBGJ:PFG6:3GK7:HYLY:5AWW:TBQ3:AROJ:OZRJ:3LZP:RMI5
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: isaal01
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8

```
(paste your output here)
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
This is Docker running on Centos 7.2 in a VMWare 6 environment.  The VM 2 vCPUs, 8GB of vRAM, and plenty of disk.  The underlying physical host has 12 x 3.3GHz CPUs, and 98 GB of RAM.  Both physical and virtual CPU usage is relatively low.

**Comment by thaJeztah** [comment]
ping @mlaventure

**Comment by mlaventure** [comment]
@AlecIsaacson could you please try with 1.12.2-rc1? 

Given that you're using exec, I'm assuming it is leaving orphaned children behind.

**Comment by AlecIsaacson** [comment]
1.12.2-rc1 is passing my basic tests, so things are looking good.  I'll have better results in a day or so.

**Comment by mlaventure** [comment]
@AlecIsaacson it's different. But still related to zombies :)

And thanks for trying the rc1!

**Comment by AlecIsaacson** [comment]
Thanks for working on this project, I really like what it has to offer.

Where can I find more details on the nature of the problem I ran into?  I’m curious about it.

**Comment by mlaventure** [comment]
@AlecIsaacson 

Actually, re-reading your details, yours is not related to zombies (it was a different case). Here is what most likely happened:

The exec into docker started a new process then exited. The process inherited the terminal IO and keep them open.

`containerd` on its side is waiting for those file descriptor to be closed to ensure no data is lost, while it does so, it will not process any new request/event which lead to the hang behavior you've observed.

 One way to reproduce the case is to do the following:

```
1$ docker run -tid --name test-exec-hang alpine top
0375d1f85814be274e969ddd5fd4ee5b1a02d819e9c1936e60d61eb4995a6173
1$ docker exec -ti test-exec-hang sh
/ # sleep 100 &
/ # exit

2$ docker run --rm -ti alpine ls
<will not execute until the above sleep 100& returns or die>
```

**Comment by AlexJF** [comment]
We just upgraded our 1.6 dockers to 1.12.1 and seem to have hit this bug. Unfortunately, I still see this with version 1.12.2-rc1.

Docker info:
```
Containers: 20
 Running: 20
 Paused: 0
 Stopped: 0
Images: 437
Server Version: 1.12.2-rc1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 506
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 3.13.0-36-generic
Operating System: Ubuntu 14.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.67 GiB
Name: jclouds-mdts-ubuntu14-testenv-da0
ID: EVA4:3MP3:ZOXC:ZJPP:XOPK:2JXG:G2NT:346F:EPYO:D4VS:5BQ4:TWVH
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 artifactory-v2.bcn.midokura.com
 127.0.0.0/8
```

Docker version:
```
Client:
 Version:      1.12.2-rc1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   45bed2c
 Built:        Tue Sep 27 21:20:26 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.2-rc1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   45bed2c
 Built:        Tue Sep 27 21:20:26 2016
 OS/Arch:      linux/amd64
```

strace to dockerd:
```
Process 8358 attached
wait4(8370,
```

strace to containerd:
```
Process 8370 attached
futex(0xee34c8, FUTEX_WAIT, 0, NULL
```

Dump from SIGUSR1 on dockerd: [docker-dump.txt](https://github.com/docker/docker/files/503251/docker-dump.txt)

Version 1.12.2-rc1 did appear to survive longer than 1.12.1 but I can't say for sure if it wasn't just some random variance. I'll try downgrading to previous versions to see if any works. It worked without problems on 1.6 though.

**Comment by AlexJF** [comment]
A small update: going back to 1.11.2, we haven't seen any hangs yet. I've also retested 1.12.1 and 1.12.2 to make sure the problem wasn't caused by something else and I've experienced the exact same hangs again.

**Comment by mlaventure** [comment]
@AlexJF do you still have the docker daemon  log that came before the stacktrace you provided earlier?

**Comment by mlaventure** [comment]
@AlexJF if I'm reading the stacktrace  correctly, the exec is waiting for either the stdout or stderr pipe to be closed.

Could you share the command of the exec that was running? If it spawned a long running process before exiting it is normal for the exec to hang there as long as that process hold onto the initial IOs.

**Comment by AlexJF** [comment]
Hi @mlaventure. Just did another run to get you the logs you requested. Also added the -D flag to the daemon:

[docker.txt](https://github.com/docker/docker/files/507934/docker.txt)

It seems to get stuck while execing `service elasticsearch restart` (from elasticsearch 2.4.0 https://github.com/elastic/elasticsearch/blob/2.4/distribution/deb/src/main/packaging/init.d/elasticsearch)

Although now there was no broken pipe error so maybe that was unrelated.

**Comment by mlaventure** [comment]
@AlexJF is that using the official elasticsearch image from the hub?

**Comment by AlexJF** [comment]
No, we are using our own image here. This is the dockerfile:

```
FROM ubuntu-upstart:14.04

# Add required packages
RUN apt-get install --no-install-recommends -y python-pip

# Add apt configuration files
RUN apt-get update -y
RUN apt-get install --no-install-recommends -y software-properties-common
RUN add-apt-repository -y ppa:openjdk-r/ppa
RUN apt-get update -y

# Add required packages
RUN apt-get install --no-install-recommends -y openjdk-8-jdk uuid-runtime curl

# Install run command
ADD bin/run.sh /run.sh
RUN chmod a+rx /run.sh

# Install Elasticsearch keys
RUN wget -qO - https://packages.elastic.co/GPG-KEY-elasticsearch | sudo apt-key add -

ONBUILD ADD conf/elasticsearch.list /etc/apt/sources.list.d/elasticsearch.list
ONBUILD RUN apt-get install -y apt-transport-https
ONBUILD RUN apt-get -q update
ONBUILD RUN apt-get install -y logstash elasticsearch
ONBUILD RUN /opt/logstash/bin/logstash-plugin install --version 0.1.2 --no-verify logstash-codec-protobuf
ONBUILD ADD conf/curator.version /etc/elasticsearch/curator.version
ONBUILD RUN pip install -U elasticsearch-curator$(cat /etc/elasticsearch/curator.version)
ONBUILD RUN echo 'ES_HEAP_SIZE="256m"' | tee -a /etc/default/elasticsearch
ONBUILD RUN echo 'LS_HEAP_SIZE="256m"' | tee -a /etc/default/logstash
ONBUILD RUN update-rc.d logstash defaults 95 10
ONBUILD RUN update-rc.d elasticsearch defaults 95 10

# Set permissions for the log file
RUN [ -f /var/log/logstash/logstash.log ] || mkdir -p /var/log/logstash
RUN touch /var/log/logstash/logstash.log
RUN chmod a+rw /var/log/logstash/logstash.log

# Install run command
ADD bin/run.sh /run.sh
RUN chmod a+rx /run.sh

# Run start script by default
CMD ["/run.sh"]
```

With run.sh being:

```
#!/usr/bin/env bash

set -ex

exec /sbin/init
```

**Comment by mlaventure** [comment]
@AlexJF thanks for the `Dockerfile` it allowed me to reproduce your issue.

It's a bug I introduced in 1.12.2, when removing the subreaper flag from `containerd` I forgot to remove the extra children wait from `containerd-shim` which mean that it is waiting for the `elasticsearch` daemonized process to exit.

You should be able to keep using docker meanwhile though, as for your `docker exec` the workaround would be to start it in the background with `-d`

I'll make a fix and add a test to avoid future regression.  Hopefully it'll get merged before `1.12.2` is released.

**Comment by thaJeztah** [comment]
fixed by https://github.com/docker/docker/issues/27160
